common/src/main/java/com/microsoft/identity/common/internal/controllers/TokenCommand.java

```diff
@@ -0,0 +1,158 @@
+
+import java.io.IOException;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+
+public class TokenCommand implements TokenOperation {
+
+    private static final String TAG = TokenCommand.class.getSimpleName();
+
+    protected OperationParameters mParameters;
+    protected BaseController mController;
+    protected List<BaseController> mControllers;
+    protected Context mContext;
+    protected IAuthenticationCallback mCallback;
+
+
+    public TokenCommand() {
+    }
+
+    public TokenCommand(@NonNull final Context context,
+                        @NonNull final OperationParameters parameters,
+                        @NonNull final BaseController controller,
+                        @NonNull final IAuthenticationCallback callback) {
+        mContext = context;
+        mParameters = parameters;
+        mController = controller;
+        mCallback = callback;
+
+        if (!(mParameters instanceof AcquireTokenSilentOperationParameters)) {
+            throw new IllegalArgumentException("Invalid operation parameters");
+        }
+    }
+
+    public TokenCommand(@NonNull final Context context,
+                        @NonNull final OperationParameters parameters,
+                        @NonNull final List<BaseController> controllers,
+                        @NonNull final IAuthenticationCallback callback) {
+        mContext = context;
+        mParameters = parameters;
+        mController = null;
+        mControllers = controllers;
+        mCallback = callback;
+
+        if (!(mParameters instanceof AcquireTokenSilentOperationParameters)) {
+            throw new IllegalArgumentException("Invalid operation parameters");
+        }
+    }
+
+    @Override
+    public AcquireTokenResult execute() throws InterruptedException, ExecutionException, IOException, ClientException, UiRequiredException, ArgumentException {
+        AcquireTokenResult result = null;
+        final String methodName = ":execute";
+
+        for(BaseController controller : mControllers) {
+            try {
+                com.microsoft.identity.common.internal.logging.Logger.verbose(
+                        TAG + methodName,
+                        "Executing with controller: " + controller.getClass().getSimpleName()
+                );
+                result = controller.acquireTokenSilent((AcquireTokenSilentOperationParameters) getParameters());
+                if(result.getSucceeded()){
+                    com.microsoft.identity.common.internal.logging.Logger.verbose(
+                            TAG + methodName,
+                            "Executing with controller: " + controller.getClass().getSimpleName() + ": Succeeded"
+                    );
+                    return result;
+                }
+            }catch(UiRequiredException e){
+                if(e.getErrorCode() == UiRequiredException.INVALID_GRANT){
+                    continue;
```

Comment:
Could you clarify the reason why we continue the acquireTokenSilent when getting INVALID_GRANT here?

Reply:
Yes, since we are iterating over a list of controllers, we will continue and try silent call with next controller in this case. I just saw there was a bug here with `equals` vs `==` for String here. Fixed it.